### PR TITLE
ROS#139: HAL initialization refactor

### PR DESCRIPTION
**Pull request by SteelPh0enix**
Refactored internal code of HAL initialization, moved peripherals to global HAL instance, changed `SystemHal` API to stateless one.

### CODE REVIEW

File: arch/cortex-m/samv71-hal/src/error.rs [hunk 1]
```diff
@@ -3,10 +3,10 @@
...
+    /// Error indicating that the requested operation was called before HAL creation.
+    HalNotCreated,
+    /// Error indicating that HAL was tried to be created twice.
     HalAlreadyCreated,
-    /// Error indicating that HAL has already been configured.
-    HalAlreadyConfigured,
-    /// Error indicating that the requested operation was called before HAL initialization.
-    HalNotInitializedYet,
+    /// Error indicating that system was tried to be initialized twice.
+    SystemAlreadyInitialized,
```
Comment (Glamhoth) [inline arch/cortex-m/samv71-hal/src/error.rs:11]: Shouldn't this be `HalAlreadyInitialized`?

File: arch/cortex-m/samv71-hal/src/hal.rs [hunk 2]
```diff
@@ -20,88 +19,85 @@ use crate::user_peripherals::UserPeripherals;
...
+static HAL_SYSTEM_PERIPHERALS: InternalCell<Option<SystemPeripherals>> = InternalCell::new(None);
 
 /// HAL implementation for Cortex-M based SAMV71 MCU.
-pub struct Hal {
-    /// User-accessible peripherals.
-    user_peripherals: Option<UserPeripherals>,
-    /// System peripherals.
-    system_peripherals: InternalCell<SystemPeripherals>,
-}
+pub struct Hal {}
```
Comment (Glamhoth) [inline arch/cortex-m/samv71-hal/src/hal.rs:30]:
```suggestion
pub struct Hal;
```

File: arch/cortex-m/samv71-hal/src/hal.rs [hunk 3]
```diff
@@ -20,88 +19,85 @@ use crate::user_peripherals::UserPeripherals;
...
+    /// the required peripherals will be missing.
     ///
-    /// # Return
-    /// [`UserPeripherals`] on first call, `None` on subsequent calls.
-    pub fn user_peripherals(&mut self) -> Option<UserPeripherals> {
-        self.user_peripherals.take()
+    /// # Returns
+    /// [`Some(UserPeripherals)`] if called for the first time after HAL initialization,
+    /// [`None`] otherwise.
+    pub fn create_user_peripherals() -> Option<UserPeripherals> {
```
Comment (Glamhoth) [inline arch/cortex-m/samv71-hal/src/hal.rs:80]: Move public functions above the private ones

File: arch/cortex-m/samv71-hal/src/hal.rs [hunk 4]
```diff
@@ -20,88 +19,85 @@ use crate::user_peripherals::UserPeripherals;
...
-}
+pub struct Hal {}
 
 impl Hal {
-    /// Frequency for the time types (TODO)
+    /// Frequency of system timer.
     const TIMER_FREQ: u32 = 1_000_000;
 
-    /// Create new HAL instance from PAC peripherals.
+    /// Create system peripherals of HAL.
```
Comment (Glamhoth) [inline arch/cortex-m/samv71-hal/src/hal.rs:36]:
```suggestion
    /// Creates system peripherals of HAL.
```

File: arch/cortex-m/samv71-hal/src/hal.rs [hunk 5]
```diff
@@ -110,49 +106,116 @@ impl SystemHal for Hal {
     type Duration = crate::time::TimerDurationU64<{ Hal::TIMER_FREQ }>;
     type Error = HalError;
 
-    fn configure_hardware(&mut self, config: SystemHardwareConfig) -> Result<(), HalError> {
-        // SAFETY: This is safe, because this is a single-core system,
-        // and no other references to system peripherals should exist.
-        let peripherals = unsafe { self.system_peripherals.as_mut_ref() };
+    /// Initialize global HAL instance using PAC peripherals.
```
Comment (Glamhoth) [inline arch/cortex-m/samv71-hal/src/hal.rs:109]:
```suggestion
    /// Initializes global HAL instance using PAC peripherals.
```

File: arch/cortex-m/samv71-hal/src/hal.rs [hunk 6]
```diff
@@ -110,49 +106,116 @@ impl SystemHal for Hal {
...
+    ///
+    /// This function executes in critical section, as it modifies HAL_SYSTEM_PERIPHERALS.
+    ///
+    /// # Safety
+    /// This function is safe to call only once.
+    /// Subsequent calls will return an error, indicating that HAL instance has already been created.
+    ///
+    /// # Return
+    /// `()` on success, [`HalError::HalAlreadyCreated`] if called more than once.
+    fn create() -> Result<(), HalError> {
```
Comment (Glamhoth) [inline arch/cortex-m/samv71-hal/src/hal.rs:123]: Is this necessary to have `fn create` and `fn configure_hardware` separate? If one has to be called after another, and `create` doesn't return anything it seems unnecessary.

File: arch/cortex-m/samv71-hal/src/system_peripherals.rs [hunk 7]
```diff
@@ -8,7 +8,8 @@ use crate::drivers::{
 };
 
 /// System peripherals structure. These peripherals are represented as HAL drivers.
-/// They are initialized on system init, and used directly by HAL to provide core functionality.
+/// Some of these peripherals are available only during HAL initialization
+/// (between Hal::create and SystemHal::configure_hardware calls).
```
Comment (Glamhoth) [inline arch/cortex-m/samv71-hal/src/system_peripherals.rs:12]:
```suggestion
/// (between SystemHal::create and SystemHal::configure_hardware calls).
```

File: src/aerugo.rs [hunk 8]
```diff
@@ -50,10 +49,7 @@ static TIME_MANAGER: TimeManager = TimeManager::new();
 /// This shouldn't be created by hand by the user or anywhere else in the code.
 /// It should be used as a [singleton](crate::aerugo::AERUGO) that acts as a system API,
 /// both for user and for the internal system parts.
-pub struct Aerugo {
-    /// Hardware Access/Abstraction Layer.
-    hal: InternalCell<Option<Hal>>,
-}
+pub struct Aerugo {}
```
Comment (Glamhoth) [inline src/aerugo.rs:52]:
```suggestion
pub struct Aerugo;
```

**Review by Glamhoth — CHANGES_REQUESTED** [review]

File: aerugo-hal/src/system_hal.rs [hunk 9]
```diff
@@ -22,19 +22,23 @@ pub trait SystemHal {
     /// Type for system HAL error.
     type Error;
 
+    /// Creates global HAL instance. Since there can only be a single instance of HAL, this function
+    /// should initialize it's global state and prepare the environment for hardware configuration.
+    fn create() -> Result<(), Self::Error>;
```
[the remark this answers is not in the document]
Comment (SteelPh0enix) [reply aerugo-hal/src/system_hal.rs:27]: yeah, good idea - renamed

File: arch/cortex-m/samv71-hal/src/hal.rs (hunk 2 above)
Comment (SteelPh0enix) [reply arch/cortex-m/samv71-hal/src/hal.rs:30]: fix'd

File: arch/cortex-m/samv71-hal/src/system_peripherals.rs (hunk 7 above)
Comment (SteelPh0enix) [reply arch/cortex-m/samv71-hal/src/system_peripherals.rs:12]: fixed, along with other typos in docs

File: src/aerugo.rs (hunk 8 above)
Comment (SteelPh0enix) [reply src/aerugo.rs:52]: fix'd

File: arch/cortex-m/samv71-hal/src/hal.rs (hunk 4 above)
Comment (SteelPh0enix) [reply arch/cortex-m/samv71-hal/src/hal.rs:36]: fix'd

File: arch/cortex-m/samv71-hal/src/hal.rs (hunk 6 above)
Comment (SteelPh0enix) [reply arch/cortex-m/samv71-hal/src/hal.rs]: You're right, i've refactored this into a single function

**Review by Glamhoth — APPROVED** [review]

**Pull request merged**